datasets/flwr_datasets/partitioner/grouped_natural_id_partitioner.py

```diff
@@ -0,0 +1,213 @@
+        self._mode = mode
+        self._sort_unique_ids = sort_unique_ids
+
+        if group_size < 0:
+            raise ValueError("group_size must be a positive integer")
+        self._group_size = group_size
+
+    def _create_int_partition_id_to_natural_id(self) -> None:
+        """Create a mapping from int indices to unique client ids from dataset.
+
+        Natural ids come from the column specified in `partition_by`.
+        """
+        unique_natural_ids = self.dataset.unique(self._partition_by)
+        if self._sort_unique_ids:
+            unique_natural_ids = sorted(unique_natural_ids)
+        num_unique_natural_ids = len(unique_natural_ids)
+        remainder = num_unique_natural_ids % self._group_size
+
+        if self._mode == "allow-bigger":
+            num_groups = num_unique_natural_ids // self._group_size
+            groups_of_natural_ids = np.array_split(unique_natural_ids, num_groups)
+        elif self._mode == "drop-reminder":
+            num_groups = num_unique_natural_ids // self._group_size
+            # Narrow down the unique_natural_ids to not have a bigger group
+            # which is the behavior of the np.array_split
+            unique_natural_ids = unique_natural_ids[
+                : int(num_groups * self._group_size)
+            ]
+            groups_of_natural_ids = np.array_split(unique_natural_ids, num_groups)
+        elif self._mode == "allow-smaller":
+            num_groups = num_unique_natural_ids // self._group_size
+            if remainder > 0:
+                last_group_ids = unique_natural_ids[-remainder:]
+            unique_natural_ids = unique_natural_ids[
+                : int(num_groups * self._group_size)
+            ]
+            groups_of_natural_ids = np.array_split(unique_natural_ids, num_groups)
+            if remainder > 0:
+                groups_of_natural_ids.append(np.array(last_group_ids))
+        elif self._mode == "strict":
+            if remainder != 0:
+                raise ValueError(
+                    "Strict mode requires that the number of unique natural ids is "
+                    "perfectly divisible by the group size. "
+                    f"Found remainder: {remainder}. Please pass the group_size that "
+                    f"enables strict mode or relax the mode parameter. Refer to the "
+                    f"documentation of the mode parameter for the available modes."
+                )
+            num_groups = num_unique_natural_ids // self._group_size
+            groups_of_natural_ids = np.array_split(unique_natural_ids, num_groups)
+        else:
+            raise ValueError(
+                f"Given {self._mode} is not a valid mode. Refer to the documentation of"
+                " the mode parameter for the available modes."
+            )
+
+        self._partition_id_to_natural_ids = {}
+        for group_of_natural_ids_id, group_of_natural_ids in zip(
+            range(len(groups_of_natural_ids)), groups_of_natural_ids
```

Comment:
is this eq to `enumerate(groups_of_natural_ids)`?

Reply:
Yes, thanks. I'll change it.